src/hotspot/cpu/aarch64/immediate_aarch64.cpp

```diff
@@ -148,11 +161,74 @@ uint64_t replicate(uint64_t bits, int nbits, int count)
   return result;
 }
 
-// this function writes the supplied bimm reference and returns a
-// boolean to indicate success (1) or fail (0) because an illegal
-// encoding must be treated as an UNALLOC instruction
+// construct a 64 bit immediate value for a logical immediate operation
+//
+// SPEC:
+//
+// {(0,_), (1, uint64)} = expandLogicalImmediate(immN, immr, imms)
+//
+// For valid combinations of immN, immr and imms, this function
+// replicates a derived bit string, whose width is a power of 2, into
+// a 64 bit result and returns 1.
+//
+// for invalid combinations it fails and returns 0
+//
+// - immN and imms together define
+//
+//    1) the size, 2^k, of the bit string to be replicated (0 < k <= 6)
+//
+//    2) the number of bits, p, to set in the string (0 < p < 2^k)
+//
+// - immr defines a right rotation on the bit string determined by
+//   immN and imms
+//
+// bit field construction:
+//
+// create a bit string of width 2^k
+//
+// set the bottom p bits to 1
+//
+// rotate the bit string right by immr bits
+//
+// replicate the 2^k bit string into 64 bits
+//
+// derivation of k and p and validity checks:
+//
+// when immN is 1 then k == 6 and immr/imms are masked to 6 bit
```

Comment:
There's a comment below "looks like 7 given the spec above but this cannot be!" - which spec is that referring to?

Reply:
I suspect that was a reference to the spec provided in an early draft of the ARM ARM -- a spec, note, that this code appears to be trying to implement in the most convoluted way possible.

I have made no attempt to update the code here. If I were to change that comment I would also be very tempted to change the code wholesale. Tweaking just that one line might suggest that the code has been updated somehow to be in line with the doc change which would be even more confusing. I think it is best to let sleeping dogs lie.